domains/client/domain-executor/src/bundle_producer.rs

```diff
@@ -311,6 +247,83 @@ where
         Ok(None)
     }
 
+    async fn propose_bundle_at(
+        &self,
+        slot: Slot,
+        primary_hash: PBlock::Hash,
+    ) -> sp_blockchain::Result<Bundle<Block::Extrinsic, NumberFor<PBlock>, PBlock::Hash, Block::Hash>>
+    {
+        let parent_number = self.client.info().best_number;
+
+        let mut t1 = self.transaction_pool.ready_at(parent_number).fuse();
+        // TODO: proper timeout
+        let mut t2 = futures_timer::Delay::new(time::Duration::from_micros(100)).fuse();
+
+        let pending_iterator = select! {
+            res = t1 => res,
+            _ = t2 => {
+                tracing::warn!(
+                    target: LOG_TARGET,
+                    "Timeout fired waiting for transaction pool at #{}, proceeding with production.",
+                    parent_number,
+                );
+                self.transaction_pool.ready()
+            }
+        };
+
+        // TODO: proper deadline
+        let pushing_duration = time::Duration::from_micros(500);
+
+        let start = time::Instant::now();
+
+        // TODO: Select transactions properly from the transaction pool
+        //
+        // Selection policy:
+        // - minimize the transaction equivocation.
+        // - maximize the executor computation power.
+        let mut extrinsics = Vec::new();
+
+        for pending_tx in pending_iterator {
+            if start.elapsed() >= pushing_duration {
+                break;
+            }
+            let pending_tx_data = pending_tx.data().clone();
+            extrinsics.push(pending_tx_data);
+        }
+
+        let extrinsics_root = BlakeTwo256::ordered_trie_root(
+            extrinsics.iter().map(|xt| xt.encode()).collect(),
+            sp_core::storage::StateVersion::V1,
+        );
+
+        let _state_root = self
+            .client
+            .expect_header(BlockId::Number(parent_number))?
+            .state_root();
```

Comment:
BTW, why is this needed (I mean extracting state root)?

Reply:
Probably useless from the legacy code, can be removed later :(